### ISSUE TITLE
Generic base class not working

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3679]
Hi!

I know that having a generic class marked with the database-attribute doesn't work. But why doesn't it work to have a non persistent generic base class?

The weaver tries to find the field __sc__this_handle__ on the type ItemHolder. That's a bug I guess?
src\scweaver\Starcounter.Internal.Weaver\BackingInfrastructure\TypeSpecificationEmit.cs:168
````this.ThisHandle = typeDef.FindField(TypeSpecification.ThisHandleName).Field.Translate(typeDef.Module);````

THIS IS JUST AN EXAMPLE, NO COMMENTS ON THE LOGIC PLEASE :)
Here comes the example that fails in the weaver:
````
class Program
    {
        static void Main()
        {
            Db.Transact(() =>
                {
                    var item = new DbItem { Name = "Std" };
                    var spcItem = new DbItemSpecial { Name = "Special" };
                    var itemHolder = new ItemHolder(item);
                    var spcItemHolder = new SpecializedItemHolder(spcItem);
                });
        }
    }

    public abstract class Template<T>
        where T : DbItem
    {
        protected Template(T item)
        {
            Item = item;
        }

        public readonly T Item;
    }

    [Database]
    public class ItemHolder : Template<DbItem>
    {
        public ItemHolder(DbItem item)
            : base(item)
        {

        }
    }

    [Database]
    public class SpecializedItemHolder : Template<DbItemSpecial>
    {
        public SpecializedItemHolder(DbItemSpecial special)
            : base(special)
        {

        }
    }

    [Database]
    public class DbItem
    {
        public string Name;
    }

    public class DbItemSpecial : DbItem
    {
        public string Surname;
    }
````

**Comment by per-samuelsson** [comment]
> But why doesn't it work to have a non persistent generic base class?

I guess the simple answer is: because we never spent any time trying to get it to work (the weaver is older then C# generics).

I guess it *could* work, but I think it will involve us to run a project for this specific thing. Remember we do quite some rewriting of classes at runtime, involving manipulating all constructor graphs, and it's already non-trivial. Involving generics won't make it easier.

**Comment by per-samuelsson** [comment]
I'm not even sure we even support non-database base classes officially. I think it's one of the things that might work in some cases, and fail in other.

In some future version, I guess we can be more forgiving to using generics in database classes, but in the current version, I think the right thing for us to do here is to raise an error saying you can't have a non-database base class.

**Comment by careri** [comment]
You're right. Doesn't work with non persistent base class. That's a set back :(
Though it opened up for such nice solutions.

Any idea why that isn't possible? I can understand that the base class can't contain database fields. So my example here is actually quite bad.

But fetching non db fields and methods from the base class should work IMHO
